pandas/core/arrays/arrow/array.py

```diff
@@ -949,7 +951,75 @@ def _indexing_key_to_indices(
             indices = np.arange(n)[key]
         return indices
 
-    # TODO: redefine _rank using pc.rank with pyarrow 9.0
+    def _rank(
+        self,
+        *,
+        axis: AxisInt = 0,
+        method: str = "average",
+        na_option: str = "keep",
+        ascending: bool = True,
+        pct: bool = False,
+    ):
+        """
+        See Series.rank.__doc__.
+        """
+        if pa_version_under9p0:
+            ranked = super()._rank(
+                axis=axis,
+                method=method,
+                na_option=na_option,
+                ascending=ascending,
+                pct=pct,
+            )
+            # keep dtypes consistent with the implementation below
+            if method == "average" or pct:
+                pa_type = pa.float64()
+            else:
+                pa_type = pa.uint64()
+            result = pa.array(ranked, type=pa_type, from_pandas=True)
+            return type(self)(result)
+
+        if axis != 0:
+            raise NotImplementedError
+
+        data = self._data.combine_chunks()
+        sort_keys = "ascending" if ascending else "descending"
+        null_placement = "at_start" if na_option == "top" else "at_end"
+        tiebreaker = "min" if method == "average" else method
+
+        result = pc.rank(
+            data,
+            sort_keys=sort_keys,
+            null_placement=null_placement,
+            tiebreaker=tiebreaker,
+        )
+
+        if na_option == "keep":
+            mask = pc.is_null(self._data)
+            null = pa.scalar(None, type=result.type)
+            result = pc.if_else(mask, null, result)
+
+        if method == "average":
+            result_max = pc.rank(
+                data,
+                sort_keys=sort_keys,
+                null_placement=null_placement,
+                tiebreaker="max",
+            )
+            result_max = result_max.cast(pa.float64())
```

Comment:
So these casts are to ensure `pc.divide` division semantics return floating results?

Reply:
Yes. `pc.rank` is able to operate entirely with `uint64` results but pandas includes two options that require floating results: `method="average"` and `pct=True`. 

If we don't cast the inputs, `pc.divide` will truncate the result keeping the `uint64` type:

```
In [1]: import pyarrow as pa

In [2]: import pyarrow.compute as pc

In [3]: a = pa.scalar(1, type="uint64")

In [4]: b = pa.scalar(2, type="uint64")

In [5]: pc.divide(a, b)
Out[5]: <pyarrow.UInt64Scalar: 0>
```